common/network-common/src/main/java/org/apache/spark/network/server/BlockPushNonFatalFailure.java

```diff
@@ -0,0 +1,162 @@
+   * BlockPushNonFatalFailure is recreated from the error code received from the server.
+   */
+  private ReturnCode returnCode;
+
+  public BlockPushNonFatalFailure(ByteBuffer response, String msg) {
+    super(msg);
+    this.response = response;
+  }
+
+  public BlockPushNonFatalFailure(ReturnCode returnCode, String msg) {
+    super(msg);
+    this.returnCode = returnCode;
+  }
+
+  /**
+   * Since this type of exception is used to only convey the error code, we reduce the
+   * exception initialization overhead by skipping filling the stack trace.
+   */
+  @Override
+  public synchronized Throwable fillInStackTrace() {
+    return this;
+  }
+
+  public ByteBuffer getResponse() {
+    // Ensure we do not invoke this method if response is not set
+    Preconditions.checkNotNull(response);
+    return response;
+  }
+
+  public ReturnCode getReturnCode() {
+    // Ensure we do not invoke this method if returnCode is not set
+    Preconditions.checkNotNull(returnCode);
+    return returnCode;
+  }
+
+  public enum ReturnCode {
+    /**
+     * Indicate the case of a successful merge of a pushed block.
+     */
+    SUCCESS(0),
+    /**
+     * Indicate a block to be merged arrives too late on the server side, i.e. after the
+     * corresponding shuffle has been merge finalized. When the client gets this code, it
+     * will not retry pushing the block.
+     */
+    TOO_LATE_BLOCK_PUSH(1),
+    /**
+     * Indicating the server couldn't append a block after all available attempts due to
+     * collision with other blocks belonging to the same shuffle partition.
+     */
+    BLOCK_APPEND_COLLISION_DETECTED(2),
+    /**
+     * Indicate a block received on the server side is a stale block push in the case of
+     * indeterminate stage retries. When the client receives this code, it will not retry
+     * pushing the block.
+     */
+    STALE_BLOCK_PUSH(3);
+
+    private final byte id;
+
+    ReturnCode(int id) {
+      assert id < 128 : "Cannot have more than 128 block push return code";
+      this.id = (byte) id;
+    }
+
+    public byte id() { return id; }
+  }
+
+  public static ReturnCode getReturnCode(byte id) {
+    switch (id) {
+      case 0: return ReturnCode.SUCCESS;
+      case 1: return ReturnCode.TOO_LATE_BLOCK_PUSH;
+      case 2: return ReturnCode.BLOCK_APPEND_COLLISION_DETECTED;
+      case 3: return ReturnCode.STALE_BLOCK_PUSH;
+      default: throw new IllegalArgumentException("Unknown block push return code: " + id);
+    }
+  }
+
+  public static String getErrorMsg(String blockId, ReturnCode errorCode) {
+    Preconditions.checkArgument(errorCode != ReturnCode.SUCCESS);
+    switch (errorCode) {
+      case STALE_BLOCK_PUSH:
+        return "Block " + blockId + STALE_BLOCK_PUSH_MESSAGE_SUFFIX;
+      case TOO_LATE_BLOCK_PUSH:
+        return "Block " + blockId + TOO_LATE_BLOCK_PUSH_MESSAGE_SUFFIX;
+      case BLOCK_APPEND_COLLISION_DETECTED:
+        return "Block " + blockId + BLOCK_APPEND_COLLISION_MSG_SUFFIX;
```

Comment:
if there is a 1:1 mapping from error code to message suffix, can you just move the message suffix to be a field on the enum? Then you can just do something like `errorCode.getMessage(blockId)` here, or `return "Block " + blockId + errorCode.getMessageSuffix()`

Reply:
Thanks for the suggestion. Modified the PR accordingly.